src/lewis/adapters/stream.py

```diff
@@ -37,27 +37,60 @@ class StreamHandler(asynchat.async_chat):
     def __init__(self, sock, target, stream_server):
         asynchat.async_chat.__init__(self, sock=sock)
         self.set_terminator(b(target.in_terminator))
-        self.target = target
-        self.buffer = []
+        self._readtimeout = target.readtimeout
+        self._readtimer = 0
+        self._target = target
+        self._buffer = []
 
         self._stream_server = stream_server
 
         self._set_logging_context(target)
         self.log.info('Client connected from %s:%s', *sock.getpeername())
 
+    def process(self, msec):
+        if len(self._buffer) > 0:
```

Comment:
It took me a while to understand that the read timeout can only happen when there's already data in the buffer. What about something like this?

```
if not self._buffer:
    return

self._readtimer += msec
# ... and so on
```